sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +376,312 @@ def transform(self, X):
+        X = check_array(X, accept_sparse=False, dtype=np.float64,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if (np.any(np.isinf(X))):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Check if % missing in any column > col_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.col_max_missing)):
+            raise ValueError("The following columns have, "
+                             "more than {0}% missing values: {1}"
+                             .format(self.col_max_missing*100, np.where(
+                              mask.sum(axis=0) > (X.shape[0] *
+                                                  self.col_max_missing))))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > col_max_missing
+        bad_rows = mask.sum(axis=1) > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "The following rows have more than {0}% missing values and "
+                "are not included as donor neighbors: {1}"
+                    .format(self.row_max_missing*100, np.where(bad_rows)))
+
+            # Remove rows that have more than row_max_missing % missing
+            X = X[~bad_rows, :]
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
+            raise ValueError("There are only %d samples, "
+                             "but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+
+        # Instantiate NN object, get column means, and store in statistics_
+        neigh = NearestNeighbors(n_neighbors=self.n_neighbors,
+                                 metric=self.metric)
+        self._fitted_neighbors = neigh.fit(X)
+        self.statistics_ = X_col_means
+
+        return self
+
+    def _transform(self, X, n_neighbors_new):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        n_neighbors_new : int
+            Indicates whether to pass n_neighbors or n_neighbors+1 to
+            _tranform().
+            Calling transform() automatically sets this to self.n_neighbors
+            while fit_transform() sets it to self.n_neighbors + 1.
+        """
+        # Import(s) here to avoid circular import
+        from ..neighbors.base import _get_weights
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
```

Comment:
maybe we should have an option `force_all_finite='nan-ok'` or something so that we don't need to repeat this testing for infinity.

Reply:
Well no one else will have used nan-ok!

Reply:
yeah but this is such a common check function I would not dare lol...but will do that now. however, to avoid potential confusion due to a negative-positive situation, maybe use "force_inf" or "force_inf_only" instead of "nan_ok"?

Reply:
`allow_nans` or `allow_nan` (whatever is used here) might be appropriate? I don't think `force_inf` is as clear

Reply:
(In fact I wish it didn't say "force" at all because it doesn't coerce, it rejects)

Reply:
totally agree on your last point -- not the most intuitive name

Reply:
I am about to add "allow_nans" and "allow_infs" into the check_array() docstring as valid input for this param. is that okay?

Reply:
Why `allow_infs`? I don't think we use infs in ML.

Reply:
just future-proofing, but maybe not :)

Reply:
Future proofing is making something you create resilient to breakage in the future, not making things just in case they'll be useful in the future.